emscripten.py

```diff
@@ -1480,10 +1480,27 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
     receiving = '\n'.join(['var real_' + s + ' = asm["' + s + '"]; asm["' + s + '''"] = function() {''' + runtime_assertions + '''  return real_''' + s + '''.apply(null, arguments);
 };
 ''' for s in exported_implemented_functions if s not in ['_memcpy', '_memset', 'runPostSets', '_emscripten_replace_memory', '__start_module']])
+
+  module_exports = exported_implemented_functions + function_tables(function_table_data)
+
+  # Generate an exports file that records all the exported symbols from asm.js/wasm module.
+  module_exports_suppressions = '\n'.join(['/**\n * @suppress {duplicate, undefinedVars}\n */\nvar %s;\n' % i for i in module_exports])
+  exports_file = get_configuration().get_temp_files().get('_module_exports.js')
+  shared.Settings.MODULE_EXPORTS_FILE = exports_file.name
+  exports_file.write(module_exports_suppressions.encode())
+  exports_file.close()
+
   if not shared.Settings.SWAPPABLE_ASM_MODULE:
-    receiving += ';\n'.join(['var ' + s + ' = Module["' + s + '"] = asm["' + s + '"]' for s in exported_implemented_functions + function_tables(function_table_data)])
+    if shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
+      receiving += ';\n'.join(['var ' + s + ' = Module["' + s + '"] = asm["' + s + '"]' for s in module_exports])
+      # TODO: Instead of the above line, we would like to use the two lines below for smaller size version of exports; but currently JS optimizer is wired to look for the exact above syntax when analyzing
+      # exports.
+#      receiving += ''.join(['var ' + s + ' = asm["' + s + '"];\n' for s in module_exports])
+#      receiving += 'for(var module_exported_function in asm) Module[module_exported_function] = asm[module_exported_function];\n'
+    else:
+      receiving += '(function() { for(var i in asm) this[i] = Module[i] = asm[i]; }());\n'
```

Comment:
use of `this` is worrying - you're modifying the global scope here. That would break in MODULARIZE etc., and also is not good for VM optimizations. But I see how it's hard to avoid it to really shrink the size down... hmm

Reply:
The intent is that in `MODULARIZE` mode, the `this[i]` exports would be assigned to the scope of the `MODULARIZE` module - does that not occur? If so, then I suppose we need to change the assignment to `{{{EXPORT_NAME}}}[i]` or something like that? I don't care for now if this is incompatible with MODULARIZE, that can be improved on later (we do use MODULARIZE, so eventually will need to)

Reply:
Hmm at least testing locally, I do not find the variable to leak outside the MODULARIZE scope to global scope.

Reply:
How did you test? If I build a hello world with MODULARIZE, and then replace the top line which is
```
var Module = (function() {
```
with this:
```
var inGlobalScope = 10;

var Module = (function() {

  console.log(this['inGlobalScope']);
```
then I see that `this` refers to the global scope, outside the modularize function.

We could perhaps work around that, by forcing `this`. But in general it's a bad idea, since it will break in other situations, like if the user applies some function scoping later on.

Reply:
I wonder if it may make sense to consider a more moderate approach here, like emitting
```
var name1, name2, .. , nameN;
for (var i in asm) Module[i] = asm[i];
```
That would not reduce all the size, but would still get the majority of the win here, and would avoid the global scope issue.

Reply:
You're correct - looks like my test was flawed - I compiled hello_world.c with MODULARIZE and poked on that in console, but that did not really get anything.

In `-s MODULARIZE=1` mode, we could do

```
var Module = (function(self) {
...
for(var i in asm) self[i] = Module[i] = asm[i];
...
});
```

and 

```
Module(Module);
```
or
```
Module({});
```

to route the assignments?